ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/Leaf.java

```diff
@@ -218,10 +224,47 @@ void optimize() {
      * @return A set of available network actions after this leaf.
      */
     Set<NetworkAction> bloom() {
-        return raoData.getAvailableNetworkActions()
-                .stream()
-                .filter(na -> !networkActions.contains(na))
-                .collect(Collectors.toSet());
+        SearchTreeRaoParameters searchTreeRaoParameters = raoParameters.getExtension(SearchTreeRaoParameters.class);
+        if (searchTreeRaoParameters.getSkipNetworkActionsFarFromMostLimitingElement()) {
+            List<Optional<Country>> worstCnecLocation = getMostLimitingElementLocation();
+            return raoData.getAvailableNetworkActions()
+                    .stream()
+                    .filter(na -> !networkActions.contains(na)
+                            && isNetworkActionCloseToLocations(na, worstCnecLocation, searchTreeRaoParameters.getMaxNumberOfBoundariesForSkippingNetworkActions()))
+                    .collect(Collectors.toSet());
+        } else {
+            return raoData.getAvailableNetworkActions()
+                    .stream()
+                    .filter(na -> !networkActions.contains(na))
+                    .collect(Collectors.toSet());
+        }
+    }
+
+    /**
+     * Says if a network action is close to a given set of countries, respecting the maximum number of boundaries
+     */
+    public boolean isNetworkActionCloseToLocations(NetworkAction networkAction, List<Optional<Country>> locations, int maxNumberOfBoundaries) {
+        if (locations.stream().anyMatch(country -> country.equals(Optional.empty()))) {
+            return true;
+        }
+        List<Optional<Country>> networkActionCountries = RaoUtil.getNetworkActionLocation(networkAction, raoData.getNetwork());
+        if (networkActionCountries.stream().anyMatch(country -> country.equals(Optional.empty()))) {
+            return true;
+        }
+        boolean isClose = false;
+        for (Optional<Country> location : locations) {
+            for (Optional<Country> networkActionCountry : networkActionCountries) {
+                if (location.isPresent() && networkActionCountry.isPresent()
+                        && CountryUtil.areNeighbors(location.get(), networkActionCountry.get(), maxNumberOfBoundaries, raoParameters.getPtdfBoundaries())) {
+                    isClose = true;
+                    break;
+                }
+            }
+            if (isClose) {
+                break;
+            }
```

Comment:
can be remove if you have a `return true` in the for loop

Reply:
yes